### ISSUE TITLE
Explore: efficiently listing a resource when creds are restricted to a subset of namespaces

### ISSUE THREAD

**Issue by kieron-dev** [cloudfoundry/cf-crd-explorations#71]
# Problem
The cf-shim executes k8s API requests to manipulate CF CRs using user credentials supplied by the CF CLI. If the user only has list permissions on a subset of namespaces, a shim API request to list all resources that a user has access becomes tricky to implement efficiently.

The cf-shim has no idea of the user permissions, so the first attempt may be a cluster-wide listing of the resource. If this succeeds, that is fine. However, the user is unlikely to have cluster-wide list permission on the resource.

The k8s API then only allows listing resources by namespace. A naive approach would be to iterate through the cluster namespaces (which is problematic in itself as our user probably will not have permission to list namespaces), listing the resources in each namespace. We would expect permission errors in any namespace the user has no privileges to list the resource in, but we could aggregate the successes, and this would provide the correct result.

However, such an approach is inefficient. It is O(N) complexity with the number of namespaces, and involves a k8s API call for each namespace. In a large cluster, it could be a long operation, and it puts unnecessary strain on the API server. Several users with namespaced permissions listing resources at the same time could overload the API server.

What is an efficient approach to avoid listing resources in namespaces where the user has no permissions to the resource? Bear in the mind the desired authorisation approach where this is delegated completely to k8s and the shim knows nothing about users, roles or role bindings.

# Areas to consider

## Can we determine the username?

There is no k8s command to give you the current user or group membership. We will definitely have a credential passed to the shim. This might be a JWT, in which case the username will be in one of the JWT properties. However, which property it is might be determined by the k8s configuration, and further the k8s configuration might namespace it with a prefix, which the shim doesn't know. Or it might be a client certificate, in which case the username is the CN of the certificate. It _might_ be basic auth, in which case we will have the username directly. Or finally we might have an opaque token which can only be translated into a username only by k8s making the appropriate API call on the underlying identity provider, which the shim has no information about.

So getting the username from the token is problematic. Maybe we can obtain it from the user info sent to an admission controller? Would having a custom resource dedicated to username divination be feasible. Its admission webhook could always reject requests, but the rejection message could contain the user details. This would add at least two additional HTTP calls to the process. Is this acceptable?

What other approaches will give us a username?

## Can we list namespaces a user has permissions in using k8s RBAC configuration?

Once we have a username, can we find out which namespaces we should restrict our listing to? Will this violate the principle that the shim knows nothing about roles and role bindings?

Things to consider:
* Can we list role bindings for the current user (presumably using service account credentials to have access), getting the set of namespaces the user has access to any resource in? Is this good enough, or might there bindings for all namespaces for some resource which breaks this?
* Could we improve the search above by restricting by a certain role? Or is this encapsulation breakage not tolerable?
* Could we break encapsulation further and inspect roles associated with a role binding to look for the resource in question?

## Can we utilise the orgs / spaces model (if it exists)?

There is talk of representing orgs and spaces as CRs. Part of that design might involve listing usernames in the org and space CRs, maybe with role details. That would be used by controllers to automate the management of the role bindings for the users.

Can we get the details we need from that structure? Or how must orgs and spaces be structured to support this?

## What sort of caching can we employ to further reduce overhead for these operations?

Can we cache token <-> username? Tokens could last anywhere from a few minutes to a few hours, so we don't necessarily have to look up usernames repeatedly.

Can we cache user namespaces? When should we invalidate the cache?

Are apps worth caching?

 ## Any other solutions

Maybe the k8s API  _can_ do this efficiently somehow?

**Comment by gcapizzi** [comment]
Is the option of just not supporting listing across orgs/spaces still on the table?

**Comment by kieron-dev** [comment]
I suppose that is the fall back if we can't find a workable solution otherwise.

I worry about supporting org level roles like org auditor. Presumably their reason is to list stuff across namespaces.

**Comment by gcapizzi** [comment]
Other question: what happens if you try to do something like [this](https://github.com/kubernetes/client-go/issues/159):

```
client.CoreV1().Pods("").List(v1.ListOptions{})
```

But you don't have permission on all namespaces? I guess it's just going to fail?

**Comment by kieron-dev** [comment]
Yep - that will fail without cluster perms.

**Comment by kieron-dev** [comment]
There is another requirement for doing this efficiently now. See https://github.com/cloudfoundry/cf-k8s-api/pull/20.

The shim needs to build up a GUID -> Namespace/Name cache, preferably using the client's credentials. It's not going to be efficient to fill the cache if we need to trawl through 1000 permission denied calls before hitting the correct namespace.

Then again, maybe the cache could be populated with a service account that does have global permissions.

**Comment by kieron-dev** [comment]
### Can we identify the user?

The worst case is that the shim receives an opaque token. This will be the case for GKE with default gcloud authentication. So let's try that.

- [ ] Is there anyway we can get the username just using the k8s API?
- [ ] If not, can we write something to do it, and what are the performance implications?

**Comment by kieron-dev** [comment]
### Built-in functionality

There is an issue [here](https://github.com/kubernetes/kubernetes/issues/30784) requesting a `whoami` API endpoint. [This comment](https://github.com/kubernetes/kubernetes/issues/30784#issuecomment-316234461) suggests RBAC debug or audit logs might be the way forward.

#### RBAC Debug

I assume this means `kubectl auth can-i ...` with debug level turned up.

In the past, `kubectl auth can-i get pods/foo -v10` would give a reason for allowing or disallowing the request including the user name. But that is no longer the case. So unless there is another form of RBAC debugging, we must drop that approach.

#### Audit Logs

[Audit logs](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/) are set up as part of the cluster configuration. They might log to a file, or call an audit webhook. Our standard problem with managed clusters is that we can't control this. We also can't guarantee the audit level - whether we get audit events for user access or not - or how we can access those logs. So I think this is another lost cause.

**Comment by kieron-dev** [comment]
### Kubectl whoami plugin

There is a kubectl plugin in existence that attempts to solve this problem: [kubectl-whoami](https://github.com/rajatjindal/kubectl-whoami). Looking at the code, it doesn't do a very good job:

- if you use a client certificate, it says username is `kubecfg:certauth:admin`
- if you use a service account token, it errors with `unsupported auth mechanism`

However, it does manage to decode an opaque gcloud token.

It works by creating a k8s auth TokenReview resource containing the token. k8s will call its token review webhook with this token, and return the user information if successful in the status.

This means we can get a username from passed credentials as follows:

| Type | Method | Comment |
| -- | -- | -- |
| Basic auth | base64 decode header and split | |
| Client cert | Get CN from cert | |
| JWT | base64 decode middle section and json unmarshal into map[string]string | we don't know the username field or any k8s prefix added to the value |
| Opaque token | Create a TokenReview resource and read the status | Involves a couple of API calls (k8s and auth webhook) |

Given that we might be reading RoleBindings for the given username, that fact that we don't accurately know what it is in the case of JWT tokens means that although promising, this approach is not sufficient.

**Comment by kieron-dev** [comment]
Following comment in the proposal about caching, we created an example using a controller-runtime cache.

We created 100 namespaces with 10 apps in each namespace. This takes ages to create, by the way.

Then we iterated over each namespace, listing the apps in each namespace.

The first access took 100ms, presumably when the cache is initially populated, and then subsequent accesses took just 10s of microseconds.

That was when using a user with cluster-wide permissions on viewing and listing the resource.

As soon as we switch to a user without cluster-wide permissions on the resource, building the cache failed.

All we could do was to create two controller-runtime 'cluster's and clients. One with cluster-wide permissions, and the other without. We engineered the second client to share the cache from the first. Unfortunately, this breaks RBAC. The user without permissions on the namespace can see them from the cache anyway. So RBAC is implemented when populating, rather than accessing, the cache.

So it looks like a cache populated by a privileged user is not a solution. And a cache without cluster permissions just doesn't work.
